crates/bevy_pbr/src/render/utils.wgsl

```diff
@@ -17,8 +18,36 @@ fn hsv2rgb(hue: f32, saturation: f32, value: f32) -> vec3<f32> {
     return value * mix(vec3<f32>(1.0), rgb, vec3<f32>(saturation));
 }
 
-fn random1D(s: f32) -> f32 {
-    return fract(sin(s * 12.9898) * 43758.5453123);
+// Generates a random u32 in range [0, u32::MAX].
+//
+// `state` is a mutable reference to a u32 used as the seed.
+//
+// Values are generated via "white noise", with no correlation between values.
+// In shaders, you often want spatial and/or temporal correlation. Use a different RNG method for these use cases.
+//
+// https://www.pcg-random.org
+// https://www.reedbeta.com/blog/hash-functions-for-gpu-rendering
+fn rand_u(state: ptr<function, u32>) -> u32 {
+    *state = *state * 747796405u + 2891336453u;
+    let word = ((*state >> ((*state >> 28u) + 4u)) ^ *state) * 277803737u;
+    return (word >> 22u) ^ word;
+}
+
+// Generates a random f32 in range [0, 1.0].
+fn rand_f(state: ptr<function, u32>) -> f32 {
+    *state = *state * 747796405u + 2891336453u;
+    let word = ((*state >> ((*state >> 28u) + 4u)) ^ *state) * 277803737u;
+    return f32((word >> 22u) ^ word) * bitcast<f32>(0x2f800004u);
+}
+
+// Generates a random vec2<f32> where each value is in range [0, 1.0].
+fn rand_vec2f(state: ptr<function, u32>) -> vec2<f32> {
+    return vec2(rand_f(state), rand_f(state));
```

Comment:
Since `rand_f` is deterministic and the same state is passed, won't this vector have the same value on both x and y?

Reply:
No because we're not passing the value of state, but the pointer itself, and rand_f will modify the value as part of the function.